File: packages/toolgovern/src/scoping/inheritance-enforcer.ts
/**
 * Default-deny scope inheritance.
 *
 * When a coordinator agent spawns a sub-agent, the sub-agent does NOT inherit the coordinator's
 * full access by default. Its scope is the intersection of what it requests and what the
 * coordinator itself actually has -- anything requested but not covered by the coordinator's own
 * scope is silently dropped, never silently granted. `ScopeRegistry` is the runtime home for
 * this: it records every agent's effective (granted) scope and re-checks it on every call a rule
 * evaluates, not just once at spawn time.
 */

import type { AgentScopeRecord, ScopeDeclaration, ScopeRegistryReader } from '../types.js';
import { credentialMatchesGranted, hostMatchesAllowed, isPathWithin } from '../shared/paths.js';
import { EMPTY_SCOPE } from './scope-declaration.js';

function intersectNetwork(
  coordinator: ScopeDeclaration['network'],
  requested: ScopeDeclaration['network'],
): ScopeDeclaration['network'] {
  if (coordinator === false || requested === false) return false;
  if (coordinator === true && requested === true) return true;
  const coordinatorList = coordinator === true ? null : coordinator;
  const requestedList = requested === true ? null : requested;
  if (coordinatorList === null) return requestedList ?? [];
  if (requestedList === null) return coordinatorList;
  // For each (coordinator, requested) pair that are the same host or one is a subdomain of
  // the other, grant the NARROWER of the two -- not unconditionally whichever list's entry
  // matched. Filtering only the coordinator list (the original bug) grants a sub-agent that
  // asked for a narrow host (`api.example.com`) the coordinator's much broader entry
  // (`example.com`) whenever the broader entry's domain happens to cover the narrow request.
  // Filtering only the requested list would fix that case but breaks the opposite one: a
  // sub-agent that broadly requests `example.com` while the coordinator holds both
  // `example.com` and a narrower `api.example.com` should still receive the narrower
  // `api.example.com` grant intact (it's covered by the broad request, not widened by it).
  const granted = new Set<string>();
  for (const coordHost of coordinatorList) {
    for (const reqHost of requestedList) {
      if (coordHost === reqHost) {
        granted.add(coordHost);
      } else if (hostMatchesAllowed(coordHost, reqHost)) {
        granted.add(coordHost); // coordinator's entry is the narrower (subdomain) value
      } else if (hostMatchesAllowed(reqHost, coordHost)) {
        granted.add(reqHost); // requested entry is the narrower (subdomain) value
      }
    }
  }
  return Array.from(granted);
}

function intersectFilesystem(
  coordinator: readonly string[],
  requested: readonly string[],
): readonly string[] {
  return requested.filter((reqPath) =>
    coordinator.some((coordPath) => isPathWithin(reqPath, coordPath)),
  );
}

function intersectCredentials(
  coordinator: readonly string[],
  requested: readonly string[],
): readonly string[] {
  return requested.filter((reqCred) =>
    coordinator.some((coordCred) => credentialMatchesGranted(reqCred, coordCred)),
  );
}

/**
 * True if `scope` carries no capability at all -- no network access (`false`, or an allowlist
 * with zero entries -- both mean "cannot reach any host"), no filesystem prefix, and no
 * credential. An agent whose *granted* scope is this empty cannot legitimately make any tool
 * call, regardless of what the call's arguments happen to look like. This is distinct from an
 * agent that simply has no filesystem scope but does have network/credential access -- only the
 * fully-empty grant means "zero tool capability."
 */
export function hasZeroCapability(scope: ScopeDeclaration): boolean {
  const hasNetwork =
    scope.network === true || (Array.isArray(scope.network) && scope.network.length > 0);
  return !hasNetwork && scope.filesystem.length === 0 && scope.credentials.length === 0;
}

/**
 * Pure function: given a coordinator's own effective scope and a sub-agent's requested scope,
 * returns the scope actually granted. Never returns anything the coordinator itself does not
 * have, and never grants anything the sub-agent did not explicitly request.
 */
export function computeInheritedScope(
  coordinatorScope: ScopeDeclaration,
  requestedScope: ScopeDeclaration,
): ScopeDeclaration {
  return {
    network: intersectNetwork(coordinatorScope.network, requestedScope.network),
    filesystem: intersectFilesystem(coordinatorScope.filesystem, requestedScope.filesystem),
    credentials: intersectCredentials(coordinatorScope.credentials, requestedScope.credentials),
  };
}

export interface SpawnSubAgentParams {
  readonly coordinatorId: string;
  readonly subAgentId: string;
  readonly sessionId: string;
  readonly requestedScope: ScopeDeclaration;
}

/**
 * Tracks every agent's effective (granted) scope for a governed run. Root agents register their
 * own declared scope directly; sub-agents are spawned against a coordinator and receive the
 * intersection of what they request and what their coordinator actually has.
 */
export class ScopeRegistry implements ScopeRegistryReader {
  private readonly records = new Map<string, AgentScopeRecord>();

  registerRootAgent(agentId: string, sessionId: string, scope: ScopeDeclaration): AgentScopeRecord {
    const record: AgentScopeRecord = { agentId, sessionId, grantedScope: scope };
    this.records.set(agentId, record);
    return record;
  }

  /**
   * Spawns a sub-agent under `coordinatorId`. If the coordinator has never been registered, the
   * sub-agent is granted the empty scope -- default-deny applies even when the caller forgot to
   * register the coordinator first, rather than falling back to "unrestricted."
   */
  spawnSubAgent(params: SpawnSubAgentParams): AgentScopeRecord {
    const coordinatorRecord = this.records.get(params.coordinatorId);
    const coordinatorScope = coordinatorRecord?.grantedScope ?? EMPTY_SCOPE;
    const grantedScope = computeInheritedScope(coordinatorScope, params.requestedScope);
    const record: AgentScopeRecord = {
      agentId: params.subAgentId,
      sessionId: params.sessionId,
      coordinatorId: params.coordinatorId,
      requestedScope: params.requestedScope,
      grantedScope,
    };
    this.records.set(params.subAgentId, record);
    return record;
  }

  getRecord(agentId: string): AgentScopeRecord | undefined {
    return this.records.get(agentId);
  }

  getEffectiveScope(agentId: string): ScopeDeclaration | undefined {
    return this.records.get(agentId)?.grantedScope;
  }

  has(agentId: string): boolean {
    return this.records.has(agentId);
  }

  /**
   * True if `agentId` is registered and its granted scope has zero capability (see
   * `hasZeroCapability`). An unregistered agent is not "zero capability" here -- that case is
   * covered separately by `TG05-unregistered-sub-agent`, so this only reports on agents the
   * registry actually has a grant on record for, whether that grant came out empty at spawn time
   * or was reduced to empty by a later coordinator re-registration.
   */
  isZeroCapability(agentId: string): boolean {
    const record = this.records.get(agentId);
    return record !== undefined && hasZeroCapability(record.grantedScope);
  }
}

File: python/src/toolgovern/scoping/inheritance_enforcer.py
"""Default-deny scope inheritance.

Ported from ``packages/toolgovern/src/scoping/inheritance-enforcer.ts``.

When a coordinator agent spawns a sub-agent, the sub-agent does NOT inherit the coordinator's
full access by default. Its scope is the intersection of what it requests and what the
coordinator itself actually has -- anything requested but not covered by the coordinator's own
scope is silently dropped, never silently granted. ``ScopeRegistry`` is the runtime home for
this: it records every agent's effective (granted) scope and re-checks it on every call a rule
evaluates, not just once at spawn time.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Optional

from ..shared.paths import credential_matches_granted, host_matches_allowed, is_path_within
from ..types import AgentScopeRecord, ScopeDeclaration
from .scope_declaration import EMPTY_SCOPE


def _intersect_network(coordinator, requested):
    if coordinator is False or requested is False:
        return False
    if coordinator is True and requested is True:
        return True
    coordinator_list = None if coordinator is True else coordinator
    requested_list = None if requested is True else requested
    if coordinator_list is None:
        return list(requested_list) if requested_list is not None else []
    if requested_list is None:
        return list(coordinator_list)
    return [
        host
        for host in coordinator_list
        if any(host_matches_allowed(host, req) or host_matches_allowed(req, host) for req in requested_list)
    ]


def _intersect_filesystem(coordinator, requested):
    return [
        req_path for req_path in requested if any(is_path_within(req_path, coord_path) for coord_path in coordinator)
    ]


def _intersect_credentials(coordinator, requested):
    return [
        req_cred
        for req_cred in requested
        if any(credential_matches_granted(req_cred, coord_cred) for coord_cred in coordinator)
    ]


def has_zero_capability(scope: ScopeDeclaration) -> bool:
    """True if ``scope`` carries no capability at all -- no network access (False, or an
    allowlist with zero entries -- both mean "cannot reach any host"), no filesystem prefix,
    and no credential. An agent whose *granted* scope is this empty cannot legitimately make
    any tool call, regardless of what the call's arguments happen to look like."""
    has_network = scope.network is True or (isinstance(scope.network, (list, tuple)) and len(scope.network) > 0)
    return not has_network and len(scope.filesystem) == 0 and len(scope.credentials) == 0


def compute_inherited_scope(
    coordinator_scope: ScopeDeclaration, requested_scope: ScopeDeclaration
) -> ScopeDeclaration:
    """Pure function: given a coordinator's own effective scope and a sub-agent's requested
    scope, returns the scope actually granted. Never returns anything the coordinator itself
    does not have, and never grants anything the sub-agent did not explicitly request."""
    return ScopeDeclaration(
        network=_intersect_network(coordinator_scope.network, requested_scope.network),
        filesystem=_intersect_filesystem(coordinator_scope.filesystem, requested_scope.filesystem),
        credentials=_intersect_credentials(coordinator_scope.credentials, requested_scope.credentials),
    )


@dataclass(frozen=True)
class SpawnSubAgentParams:
    coordinator_id: str
    sub_agent_id: str
    session_id: str
    requested_scope: ScopeDeclaration


class ScopeRegistry:
    """Tracks every agent's effective (granted) scope for a governed run. Root agents register
    their own declared scope directly; sub-agents are spawned against a coordinator and receive
    the intersection of what they request and what their coordinator actually has."""

    def __init__(self) -> None:
        self._records: Dict[str, AgentScopeRecord] = {}

    def register_root_agent(self, agent_id: str, session_id: str, scope: ScopeDeclaration) -> AgentScopeRecord:
        record = AgentScopeRecord(agent_id=agent_id, session_id=session_id, granted_scope=scope)
        self._records[agent_id] = record
        return record

    def spawn_sub_agent(self, params: SpawnSubAgentParams) -> AgentScopeRecord:
        """Spawns a sub-agent under ``params.coordinator_id``. If the coordinator has never
        been registered, the sub-agent is granted the empty scope -- default-deny applies even
        when the caller forgot to register the coordinator first, rather than falling back to
        "unrestricted."."""
        coordinator_record = self._records.get(params.coordinator_id)
        coordinator_scope = coordinator_record.granted_scope if coordinator_record else EMPTY_SCOPE
        granted_scope = compute_inherited_scope(coordinator_scope, params.requested_scope)
        record = AgentScopeRecord(
            agent_id=params.sub_agent_id,
            session_id=params.session_id,
            coordinator_id=params.coordinator_id,
            requested_scope=params.requested_scope,
            granted_scope=granted_scope,
        )
        self._records[params.sub_agent_id] = record
        return record

    def get_record(self, agent_id: str) -> Optional[AgentScopeRecord]:
        return self._records.get(agent_id)

    def get_effective_scope(self, agent_id: str) -> Optional[ScopeDeclaration]:
        record = self._records.get(agent_id)
        return record.granted_scope if record else None

    def has(self, agent_id: str) -> bool:
        return agent_id in self._records

    def is_zero_capability(self, agent_id: str) -> bool:
        """True if ``agent_id`` is registered and its granted scope has zero capability. An
        unregistered agent is not "zero capability" here -- that case is covered separately by
        TG05-unregistered-sub-agent."""
        record = self._records.get(agent_id)
        return record is not None and has_zero_capability(record.granted_scope)
